### PR TITLE
Add support for Odroid M1S Board

### PR DESCRIPTION
**Pull request by darkxst**
Hardkernel have recently released the Odroid M1S which is a slimmed down version of the original Odroid M1.

Features:
- Rockchip RK3566
- 4 or 8GB LPDDR4 RAM
- 64GB eMMC (soldered)
- Pcie 2.1 NVMe M2 slot
- new 14pin GPIO header that supports a variety of addon boards (most require DT overlays though)

This needs to be added as a new board since it uses a different TPL blob for DDR initialiasation.

There is no upstream support for this board just yet, however that is in progress. For now I am using the following:
Uboot using generic-rk3568 target
Linux device tree synced from the unofficial hardkernel 6.1 linux
https://github.com/tobetter/linux/tree/odroid-6.1.y

Installation to eMMC can be done directly with Hardkernel's USB UMS tool
https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide?redirect=1#install_over_usb_from_pc

### CODE REVIEW

**Review by sairon — COMMENTED** [review]
Although this PR is still a draft, could you please put it on hold? We plan to merge #3063 this week and it will need to be updated afterwards, so I wanted to give you a heads-up about that. I remember you wrote you developed the support also for 6.6 kernel, so I hope this is not a big complication for you.

Maybe @agners has some more comments regarding this PR as well.

But anyway, thanks for your work!

**Comment by darkxst** [comment]
> I remember you wrote you developed the support also for 6.6 kernel, so I hope this is not a big complication for you.

Yes,  I have a 6.6 branch as well, so no issues there. I will rebase onto that once 6.6 is merged.

**Comment by darkxst** [comment]
First 2 patches are actually from PR #3165, but apart from that this should be ready for review.

Changes from first round:
- Rebased onto 6.6 12.0dev branch
- Switch to full uboot config, as there were issues with ethernet and nvme after booting kernel when using generic config. This also will read the proper MAC address from /dev/mmcblk0boot0. (Based on patch from uboot lists, but it wont be merged until device tree is in linux!)
- Activate UMS gadget on OTG port,allowing to mount internal eMMC on PC, making it easy for user to reinstall etc if required. The hardkernel UMS tool wont boot once mainline uboot is installed. I've also included a serial console on the OTG at the same time.

**Review by sairon — COMMENTED** [review]
Looks good from my perspective :+1: There are just some nits below.

However I don't have M1S yet to try it out - but it should be somewhere on the way, same for @agners.

If there's no concern on Stefan's side, we could merge it before we start 12.0 beta, hopefully we should get M1S and add it to our board farm for testing before the public release.

File: buildroot-external/board/hardkernel/odroid-m1s/meta
Comment (darkxst) [reply buildroot-external/board/hardkernel/odroid-m1s/meta:11]: I think Agners said we would share same supervisor etc, its mainly only because of the TPL DDR init code differing that mandated a new board.

File: buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch
Comment (darkxst) [reply buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch:4]: It is needed to match up with `linux/0005-Remove-deprecated-snps-reset-properties.patch`

The snps,reset properties under &gmac are deprecated and moved to properties under &mdio (see above patch). However with this the kernel is unable to reset the ethernet phy correctly and requires the reset to happen within uboot. There was a long discussion upstream on linux list about chicken+egg problems with the reset, where kernel can't reset until it get id, but can't get id until reset or something. So Kwiboo fixed it via uboot resetting.

This is inline with how the Odroid-m1 is handled upstream and probably all other rk356x boards now. Green would also need these same changes at some point, prior to upstreaming.

File: buildroot-external/configs/odroid_m1s_defconfig
Comment (darkxst) [reply buildroot-external/configs/odroid_m1s_defconfig]: Indeed, have fixed now.

File: buildroot-external/board/hardkernel/odroid-m1s/meta
Comment (agners) [reply buildroot-external/board/hardkernel/odroid-m1s/meta:11]: Granted, we typically create a separate Supervisor machine for each board. However, we debated just recently if that still make sense. 

In practice the things which requires board specific Supervisor machine is if we need board specific user space software. E.g. to leverage hardware accelerators or similar. Often these things are SoC specific.

We do have very little boards specific things in the Core container today, e.g. all Raspberry Pi board containers install the same packets in Core (see https://github.com/home-assistant/core/tree/dev/machine). But then, it only creates a small board specific layer, so it isn't a big burden for the build systems etc. And it would allow to create a board specific changes, if we really want to. So we decided to keep the system in place as is.

However, since ODROID-M1 and ODROID-M1S use the same SoC family with the same NPU and multi-media engine, and the board is from the same vendor, I'd say it is not worth making the distinction here. And if we really need to, we can introduce this in a later update.

File: buildroot-external/board/hardkernel/odroid-m1s/patches/linux/0004-set-peripheral-mode-for-usb-otg-port.patch
Comment (darkxst) [reply buildroot-external/board/hardkernel/odroid-m1s/patches/linux/0004-set-peripheral-mode-for-usb-otg-port.patch:4]: I think the automatic switching requires a special OTG cable which shorts the ID pin to ground indicating to switch modes.

I suppose however there might be other ways to switch the mode without hard coding it in the DT.

**Comment by darkxst** [comment]
> So you are exporting `/dev/mmcblk0` as a block device, but that is also used for the data partition etc. I don't think it is safe to mount partitions/write to the block device when there are read-write partitions on that device 😰

Yes, that could be an issue, I was thinking it would provide an easy way for user to reinstall system etc, but even that wont necessarily be safe. Not sure if there is a way around this?

> You are currently using the kernel modules directly, from what I know this is the legacy way of enabling USB gadget modules

Yes I am aware that modules are considered legacy, however the configfs way is 100x more complicated than the legacy  modules when used directly. I will however have alook and see if some tools help with that!

**Comment by agners** [comment]
> Yes, that could be an issue, I was thinking it would provide an easy way for user to reinstall system etc, but even that wont necessarily be safe. Not sure if there is a way around this?

We could introduce a separate systemd target which stops before mounting the data partition, but instead enables USB gadget mode (e.g. `haos-recovery.target`). Altough, I guess it would still not be safe as the OS might read things from the squashfs still :cry: 

I guess the only way to do this safe would be a rootfs/squashfs we load into RAM at boot. Or pivot into a temporary root. That is what OpenWrt does IIRC.

In any case, I'd suggest to split out/revert the USB gadget stuff from this PR so we can get ODROID-M1S going.

> Yes I am aware that modules are considered legacy, however the configfs way is 100x more complicated than the legacy modules when used directly. I will however have alook and see if some tools help with that!

Yeah tools make it easier, but it is still more complicated. But then, also more flexible :sweat_smile:

**Comment by sairon** [comment]
I agree with Stefan, let's separate the basic board support from the USB gadget stuff, as it may get fairly complex. Then, and after fixing merge conflicts caused by #3165 merge, we can move on with merging the M1S support 🎉

**Comment by darkxst** [comment]
> We could introduce a separate systemd target which stops before mounting the data partition, but instead enables USB gadget mode (e.g. `haos-recovery.target`). 
> 
> I guess the only way to do this safe would be a rootfs/squashfs we load into RAM at boot. Or pivot into a temporary root. That is what OpenWrt does IIRC.

Yeh either of those could work.

I have rebased on dev and split out the UMS, will create a separate PR for it.

File: buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch
Comment (agners) [reply buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch:4]: Is that maybe also related to what I've seen on my ODROID-M1? I've had to disable Ethernet in U-Boot in this case though, see #3122.

**Review by agners — APPROVED** [review]

File: buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch
Comment (darkxst) [reply buildroot-external/board/hardkernel/odroid-m1s/patches/uboot/0004-Fix-ethernet-properties.patch:4]: I suppose it could be related, odroid M1 is already using the new properties though (atleast in 6.6 / 2024.01). However I can't actually find the upstream thread discussing this now.

**Review by sairon — APPROVED** [review]
Looks good, thank you for your work :+1:

**Pull request merged**
